feat(command): ng test command runs karma

Brocco opened the pull request:
Overrode the ember-cli test command to initialize the karma server
Then start the karma server based upon the karma.conf.js config file

closes #70

cironunes commented on addon/ng2/commands/test.js:
```diff
@@ -0,0 +1,32 @@
+    },
+
+    run: function(commandOptions, rawArgs) {
+      var promise = new Promise(function(){});
```
What about having `this.server.start()` inside this function?

```js
var promise = new Promise(function() {
  return this.server.start();
});
```

Brocco commented:
That works update coming soon.

Been doing a whole bunch of TypeScript recently, so it didn't occur to me that lexical scoping via fat arrow wasn't in play in your snippet, so it had to be tweaked to:

```
var server = this.server;

return new Promise(function(){
  return server.start();
});
```

IgorMinar commented on addon/ng2/commands/test.js:
```diff
@@ -0,0 +1,30 @@
+
+    init: function(){
+      var cwd = process.cwd();
+      this.karma = this.karma || require(cwd + '/node_modules/karma/lib/index');
```
use `require.resolve` instead

Brocco replied:
@IgorMinar I'm not familiar w/ `require.resolve` are you suggesting that I use do this:

`this.karma = this.karma || require(require.resolve(cwd + '/node_modules/karma/lib/index'));`

IgorMinar replied:
no that's not right. the problem is that we can't be hardcoding paths because it will cause issues on windows.

why can't we just say `require('karma')`? wouldn't that resolve correctly?

`cwd` points to the application dir and this file should be executed from `cwd` + '/node_modules/angular-cli/addon/ng2/commands/test.js` no? If that's the case then `require('karma')` should just work...

IgorMinar commented:
shouldn't we trigger `ng build` when running `ng test`?

otherwise this looks good, I think. 

@cironunes can you take a look as well?

@cironunes, @Brocco can you open another issue to integrate the build pipeline with karma so that we automatically rebuild the project when `src` changes, which then updates `dist`. we should then tell karma to rerun the tests. `angular/angular` repo has this setup. the implementation is a bit messy but it's a good start.

cironunes commented:
@Brocco, nice one. As Igor said, please just make sure to run `ng build` before starting karma.

@IgorMinar, created the issue #87

Brocco commented:
ng test will now execute a build and kick off karma

IgorMinar commented on tests/acceptance/test.spec.js:
```diff
@@ -0,0 +1,53 @@
+    return ng([
+      'test',
+      '--skip-npm',
+      '--skip-bower'
```
do we need these two flags?

Brocco replied:
I went back and checked ember's implementation and couldn't find a test spec for the test command. I think it is because we would need to run an npm install... otherwise requiring karma will fail

IgorMinar replied:
but `ng init` is executed above with these flags. I don't thing we need the flags here. they don't do anything. try removing.

IgorMinar commented on tests/acceptance/test.spec.js:
```diff
@@ -0,0 +1,53 @@
+  });
+
+  afterEach(function () {
+    this.timeout(10000);
```
why are you increasing timeout here? does it take up to 10 seconds to teardown the tmp dir?

Brocco replied:
borrowed (read copied) from the new.spec.js

IgorMinar replied:
hmm.. try removing and we'll see what happens. @rodyhaddad and @cironunes set up the other tests, so I'm not sure if this stuff is actually required.

IgorMinar commented on addon/ng2/commands/test.js:
```diff
@@ -0,0 +1,59 @@
+        })
+        .then(function(){
+          return new Promise(function(){
+            return self.karmaServer.start();
```
I don't believe that `start()` returns anything. does it?

IgorMinar replied:
if it does not then the promise returned by run() will not wait until karma is done running tests

IgorMinar commented on addon/ng2/commands/test.js:
```diff
@@ -0,0 +1,59 @@
+      var self = this;
+      return win.checkWindowsElevation(this.ui)
+        .then(function() {
+          return buildTask.run(buildCommandOptions);
```
we should support passing through karma command line flags like `--single-run` into the server, otherwise I'm not sure how we'll use `ng test` on CI

Brocco replied:
@IgorMinar can you compile a list of command line flags that you'd like to see included?

I already have `--single-run` w/ `-sr` alias implemented

IgorMinar replied:
we can do one of:

- pass all arguments thought to karma (could cause collisions between cli args and karma args)
- pass through all args recognized by karma (could still cause collisions, but fewer)
- require separator between cli args and karma args, e.g. `--` is commonly used for this purpose


I think the third option is the best. example usage:

```
ng test --env=production -- --single-run --browsers=Chrome,Safari
```

cironunes replied:
we need to keep in mind that there's still the e2e tests via protractor

IgorMinar commented on tests/acceptance/test.spec.js:
```diff
@@ -0,0 +1,53 @@
+      '--skip-bower'
+    ])
+      .then(function () {
+        console.log('test completed');
```
I'm suspicious of us actually waiting until the test run is done

IgorMinar commented on tests/acceptance/test.spec.js:
```diff
@@ -0,0 +1,53 @@
+    ])
+      .then(function () {
+        console.log('test completed');
+      });
```
what shuts down the karma server? is it shut down when the mocha test is done running? this is not right because if we have two tests then they'll bring up two karma servers and the second will fail because the port is going to be taken by the first one.

Brocco replied:
The addition of the `--single-run` command line arg will handle the karma server shut down. I've moved the karma server creation into the run function (no longer in init) to allow for passing of command line args it has to be in `run`

cironunes commented:
Protractor integration is getting done: #94.

Will we integrate it on `ng test` in this PR or do you want me to create a separate issue so we can discuss?

We'll probably need some flags as Igor were saying `ng test --all(default)/unit/e2e`

IgorMinar commented:
I suggest we commit this first and then add protractor in a follow up PR

wrt `ng test` running protractor tests - I dunno, historically we've always treated e2e tests differently because they are slow and people rarely run them locally (only if CI is broken). So I think that we should make it easy to run e2e tests locally, but don't run them by default.

I'm open to be convinced otherwise. But again, we can discuss that in a follow up PR

IgorMinar commented:
@Brocco are you blocked on anything? If not, can you resolve the comments above please so that we can merge this in?

Brocco commented:
@IgorMinar I'm still not sure how to go about testing the test command (`npm test` from the cli project)

I updated this branch/PR to exclude the running of the ng test acceptance tests (`xdescribe`) in case anyone else wants to take a look in the meantime.

I added the karma args and defaulted it to be a single-run to alleviate confusion "why's that still running"

Brocco commented:
Updated the readme to update the documentation on how to execute tests

cironunes commented:
@Brocco  can you rebase this with the latest changes from master?

jkuri commented:
Signed.

> On 24 Jan 2016, at 03:17, googlebot <notifications@github.com> wrote:
> 
> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
>  Please visit https://cla.developers.google.com/ <https://cla.developers.google.com/> to sign.
> 
> Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.
> 
> If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address. Check your existing CLA data <https://cla.developers.google.com/clas> and verify that your email is set on your git commits <https://help.github.com/articles/setting-your-email-in-git/>.
> If you signed the CLA as a corporation, please let us know the company's name.
> —
> Reply to this email directly or view it on GitHub <https://github.com/angular/angular-cli/pull/86#issuecomment-174243735>.
>

Brocco commented:
I signed it!
On Jan 23, 2016 9:17 PM, "googlebot" <notifications@github.com> wrote:

> Thanks for your pull request. It looks like this may be your first
> contribution to a Google open source project. Before we can look at your
> pull request, you'll need to sign a Contributor License Agreement (CLA).
>
> [image: :memo:] *Please visit https://cla.developers.google.com/
> <https://cla.developers.google.com/> to sign.*
>
> Once you've signed, please reply here (e.g. I signed it!) and we'll
> verify. Thanks.
> ------------------------------
>
>    - If you've already signed a CLA, it's possible we don't have your
>    GitHub username or you're using a different email address. Check your
>    existing CLA data <https://cla.developers.google.com/clas> and verify
>    that your email is set on your git commits
>    <https://help.github.com/articles/setting-your-email-in-git/>.
>    - If you signed the CLA as a corporation, please let us know the
>    company's name.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angular/angular-cli/pull/86#issuecomment-174243735>.
>

cironunes commented:
@Brocco have you updated the PR?

I'm still seeing:

> This branch is out-of-date with the base branch
Merge the latest changes from master into this branch.

filipesilva commented on addon/ng2/blueprints/ng2/files/karma.conf.js:
```diff
@@ -14,6 +15,8 @@ module.exports = function(config) {
       {pattern: 'node_modules/angular2/bundles/http.dev.js', included: true, watched: true},
       {pattern: 'node_modules/angular2/bundles/router.dev.js', included: true, watched: true},
       {pattern: 'node_modules/angular2/bundles/testing.dev.js', included: true, watched: true},
+      {pattern: 'node_modules/es6-shim/es6-shim.js', included: true, watched: true},
```
Shouldn't these be in the same order as https://github.com/angular/angular-cli/blob/master/lib/broccoli/angular2-app.js#L21-L30 ?

filipesilva commented on addon/ng2/blueprints/ng2/files/package.json:
```diff
@@ -19,7 +19,7 @@
     "reflect-metadata": "0.1.2",
     "rxjs": "5.0.0-beta.0",
-    "systemjs": "0.19.4"
+    "systemjs": "0.19.20"
```
Good catch.

Brocco replied:
Thanks

filipesilva commented on addon/ng2/blueprints/route/files/src/app/__name__/__name__-detail.component.html:
```diff
@@ -1,13 +1,15 @@
-  <div>
-    <label>Id:</label>
-    {{<%= camelizedModuleName %>.id}}
+<div>
```
What is the rationale behind the enveloping extra div?

Brocco replied:
Hold over from something I tried, I will fix and rebase

filipesilva commented on tests/e2e/e2e_workflow.spec.js:
```diff
@@ -56,19 +67,17 @@ describe('Basic end-to-end Workflow', function () {
-      done();
+    return ng(testArgs)
+    .then(function(result) {
+      expect(result.exitCode).to.be.equal(0);
```
Is this what we have to check on future tests for the correct execution of a command?

Brocco replied:
for karma tests yes, based on their API it is how we can discern if our tests passed/failed via our CLI

filipesilva commented:
Other than the comments I added in, lgtm.
